pushsync/pusher.go

```diff
@@ -0,0 +1,343 @@
+
+			// and start iterating on Push index from the beginning
+			chunks, unsubscribe = p.store.SubscribePush(ctx)
+			// reset timer to go off after retryInterval
+			timer.Reset(retryInterval)
+
+		// handle incoming chunks
+		case ch, more := <-chunks:
+			// if no more, set to nil, reset timer to 0 to finalise batch immediately
+			if !more {
+				chunks = nil
+				timer.Reset(0)
+				break
+			}
+
+			chunksInBatch++
+			metrics.GetOrRegisterCounter("pusher.send-chunk", nil).Inc(1)
+			// if no need to sync this chunk then continue
+			if !p.needToSync(ch) {
+				break
+			}
+
+			metrics.GetOrRegisterCounter("pusher.send-chunk.send-to-sync", nil).Inc(1)
+			// send the chunk and ignore the error
+			// go func(ch chunk.Chunk) {
+			if err := p.sendChunkMsg(ch); err != nil {
+				log.Error("error sending chunk", "addr", ch.Address().Hex(), "err", err)
+			}
+
+		// handle incoming receipts
+		case addr := <-p.receipts:
+			hexaddr := hex.EncodeToString(addr)
+			log.Debug("synced", "addr", hexaddr)
+			metrics.GetOrRegisterCounter("pusher.receipts.all", nil).Inc(1)
+			// ignore if already received receipt
+			item, found := p.pushed[hexaddr]
+			if !found {
+				metrics.GetOrRegisterCounter("pusher.receipts.not-found", nil).Inc(1)
+				log.Debug("not wanted or already got... ignore", "addr", hexaddr)
+				break
+			}
+			if item.synced {
+				metrics.GetOrRegisterCounter("pusher.receipts.already-synced", nil).Inc(1)
+				log.Debug("just synced... ignore", "addr", hexaddr)
+				break
+			}
+
+			totalDuration := time.Since(item.firstSentAt)
+			if !item.shortcut {
+				roundtripDuration := time.Since(item.lastSentAt)
+				measurement := uint64(roundtripDuration) / 1000
+				if 2*measurement < 3*average {
+					average = (measurements*average + measurement) / (measurements + 1)
+					measurements++
+					retryInterval = time.Duration(average*2) * time.Millisecond
+					log.Debug("time to sync", "addr", hexaddr, "total duration", totalDuration, "roundtrip duration", roundtripDuration, "n", measurements, "average", average, "retry", retryInterval)
+				}
+			}
+			metrics.GetOrRegisterResettingTimer("pusher.chunk.roundtrip", nil).Update(totalDuration)
+			metrics.GetOrRegisterCounter("pusher.receipts.synced", nil).Inc(1)
+			// collect synced addresses and corresponding items to do subsequent batch operations
+			syncedAddrs = append(syncedAddrs, addr)
+			syncedItems = append(syncedItems, item)
+			// set synced flag
+			item.synced = true
+
+		case <-p.quit:
+			// if subscribe was running, stop it
+			if unsubscribe != nil {
+				unsubscribe()
+			}
+			return
+		}
+	}
+}
+
+// handleReceiptMsg is a handler for pssReceiptTopic that
+// - deserialises receiptMsg and
+// - sends the receipted address on a channel
+func (p *Pusher) handleReceiptMsg(msg []byte) error {
+	receipt, err := decodeReceiptMsg(msg)
+	if err != nil {
+		return err
+	}
+	log.Debug("Handler", "receipt", label(receipt.Addr), "self", label(p.ps.BaseAddr()))
```

Comment:
Trace

Reply:
If you need `BaseAddr` in the interface just for logging - it is better to just create a logger in `Pusher` that will just append this address to all loglines (recommended!)